CONTRIBUTING.md

```diff
@@ -18,7 +18,117 @@ These are general and loose rules components in UI Kit should strive to fulfill.
 - Documentation of prop-types and usage patterns
 - Sensible defaults for props which are not a hard requirement
 
-### How to add a component to UI Kit.
+## Development
+
+### Testing
+
+We split our testing in three parts: functionality, visual and the final bundle. We use separate tools for these, and we try to avoid overlap.
+
+#### Functionality testing
+
+- write tests from a user perspective, not from a dev perspective
+- we test using `react-testing-library`
+  - we have `test-utils` as a thin wrapper which sets up the context of tested components
+  - we avoid `beforeEach` as it leads to test coupling
+  - an `it` can describe the flow of a user
+  - it is okay to have many `expect` statements in an `it`
+  - code comments can help explain the `expect`s better
+- avoid testing for visuals (classnames or the DOM structure)
+  - the user does not care about the DOM structure or the classnames themselves
+  - use visual testing instead (see below)
+- if mocking is required, try to do it at the level the user interacts with the component
+  - Examples
+    - good: mocking network responses
+    - bad: mocking a fetching component
+  - Try to keep mocking minimal. The less we mock, and the later we mock, the more our tests will cover and the more confidence we gain that our components work
+
+#### Visual testing
+
+- we will use a tool called [Percy](https://percy.io/) for visual regression testing
+- we test the components from the produced bundle
+  - this gives us confidence that the bundling process works
+  - this tests components under conditions closer to how consumers would use them
+  - this helps us to avoid accidentally not exposing implemented components from the main `index.js` file
+- we test components in multiple states in one snapshot
+  - it is nice to see all states in one snapshot
+  - it uses less snapshots (we have a monthly budget)
+  - we use custom components called `Suite` and `Spec` to surround our component-under-test, which makes the review easier
+  - we set a min-height for every `Spec` to ensure that changes in height of a component's state don't break the visual diff of states following below that
+- we can currently not "interact" with components (like opening a dropdown menu) before taking a screenshot. We might be able to solve this by using Puppeteer in the future, but we wanted to get up and running first.
+
+#### Bundle testing
+
+We have test files ending with `*.bundlespec.js`. They are used to test the bundle produced by the rollup build. It is necessary to run the build before running these tests. We use these to further ensure that our bundling process works.
```

Comment:
```suggestion
We have test files ending with `*.bundlespec.js`. They are used to test the bundle produced by the rollup build, to ensure that our bundling process works.

Note that it is necessary to run the build before running these tests.
```

Reply:
This suggestion is missing the part which says that one needs to build before running the test. And I don't like the repeated use of "to ... to ...". So I left this one out. Thanks for the other two suggestions!

Reply:
That line is there is the suggestion, didn’t you see it? 😅